<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Menzerath's law in biology • menzerath</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Menzerath's law in biology">
<meta property="og:description" content="menzerath">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">menzerath</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/menzerath.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/menzerath_biology.html">Menzerath's law in biology</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sellisd/menzerath/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="menzerath_biology_files/accessible-code-block-0.0.1/empty-anchor.js"></script><script src="menzerath_biology_files/kePrint-0.0.1/kePrint.js"></script><link href="menzerath_biology_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Menzerath’s law in biology</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/sellisd/menzerath/blob/master/vignettes/menzerath_biology.Rmd"><code>vignettes/menzerath_biology.Rmd</code></a></small>
      <div class="hidden name"><code>menzerath_biology.Rmd</code></div>

    </div>

    
    
<p>Menzerath’s law (also known as Menzerath-Altmann law) in linguistics describes the relationship between the size of a linguistic construct (e.g. a word) and its constituents (e.g. syllables), which are expected follow the relationship:</p>
<p><span class="math display">\[ y = a \cdot x^b \cdot e^{-cx} \]</span> where <code>a</code>, <code>b</code> and <code>c</code> are parameters of the law, <code>x</code> is the size of the construct measured in units of its constituents and <code>y</code> is the average size of constituents measured in units of a subconsistuent. Menzerath’s law can apply to complex systems with multiple levels and has been found in other systems outside linguistics.</p>
<p>Menzerath’s law has been found in a number of different biological sequences and organizational levels including:</p>
<ul>
<li>Genomes, chromosomes and base pairs <span class="citation">(Ferrer-I-Cancho and Forns 2009; Baixeries et al. 2013)</span>
</li>
<li>Genes, exons and base pairs <span class="citation">(Li 2011)</span> and perhaps more appropriately transcripts, exons and base pairs <span class="citation">(Nikolaou 2014)</span>
</li>
<li>Proteins, protein domains and aminoacids <span class="citation">(Shahzad, Mittenthal, and Caetano-Anollés 2015)</span>
</li>
</ul>
<p>To facilitate the exploration of Menzerath’s law we can use the <a href="https://sellisd.github.io/menzerath/">menzerath R package</a>. Lets consider the case of the transcript-exon level as an example. We use <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> package to give us access to the the biological data. The package is part of <a href="bioconductor.org">Bioconductor</a> and provides an interface to BioMart databases such as <a href="https://www.ensembl.org/">Ensemble</a>, where we can download structural information for a number of genomes.</p>
<p>First we have to load the packages we are going to use (If any of the packages is missing check the <a href="#Install-missing-libraries-section">Install Missing Libraries section</a>):</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://tidyverse.tidyverse.org">"tidyverse"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/sellisd/menzerath">"menzerath"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://wilkelab.org/cowplot/">"cowplot"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span></code></pre></div>
<p>Once everything is in place we can start by connecting to <a href="https://www.ensembl.org/index.html">ensemble</a>, a genome database project of the European Bioinformatics Institute, and list the databases (here are the first 20):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"ensembl"</span><span class="op">)</span>
<span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/listDatasets.html">listDatasets</a></span><span class="op">(</span><span class="va">ensembl</span><span class="op">)</span>
<span class="fu">kbl</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">datasets</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr>
<th style="text-align:left;">
dataset
</th>
<th style="text-align:left;">
description
</th>
<th style="text-align:left;">
version
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
abrachyrhynchus_gene_ensembl
</td>
<td style="text-align:left;">
Pink-footed goose genes (ASM259213v1)
</td>
<td style="text-align:left;">
ASM259213v1
</td>
</tr>
<tr>
<td style="text-align:left;">
acalliptera_gene_ensembl
</td>
<td style="text-align:left;">
Eastern happy genes (fAstCal1.2)
</td>
<td style="text-align:left;">
fAstCal1.2
</td>
</tr>
<tr>
<td style="text-align:left;">
acarolinensis_gene_ensembl
</td>
<td style="text-align:left;">
Anole lizard genes (AnoCar2.0)
</td>
<td style="text-align:left;">
AnoCar2.0
</td>
</tr>
<tr>
<td style="text-align:left;">
acchrysaetos_gene_ensembl
</td>
<td style="text-align:left;">
Golden eagle genes (bAquChr1.2)
</td>
<td style="text-align:left;">
bAquChr1.2
</td>
</tr>
<tr>
<td style="text-align:left;">
acitrinellus_gene_ensembl
</td>
<td style="text-align:left;">
Midas cichlid genes (Midas_v5)
</td>
<td style="text-align:left;">
Midas_v5
</td>
</tr>
<tr>
<td style="text-align:left;">
amelanoleuca_gene_ensembl
</td>
<td style="text-align:left;">
Giant panda genes (ASM200744v2)
</td>
<td style="text-align:left;">
ASM200744v2
</td>
</tr>
<tr>
<td style="text-align:left;">
amexicanus_gene_ensembl
</td>
<td style="text-align:left;">
Mexican tetra genes (Astyanax_mexicanus-2.0)
</td>
<td style="text-align:left;">
Astyanax_mexicanus-2.0
</td>
</tr>
<tr>
<td style="text-align:left;">
anancymaae_gene_ensembl
</td>
<td style="text-align:left;">
Ma’s night monkey genes (Anan_2.0)
</td>
<td style="text-align:left;">
Anan_2.0
</td>
</tr>
<tr>
<td style="text-align:left;">
aocellaris_gene_ensembl
</td>
<td style="text-align:left;">
Clown anemonefish genes (AmpOce1.0)
</td>
<td style="text-align:left;">
AmpOce1.0
</td>
</tr>
<tr>
<td style="text-align:left;">
apercula_gene_ensembl
</td>
<td style="text-align:left;">
Orange clownfish genes (Nemo_v1)
</td>
<td style="text-align:left;">
Nemo_v1
</td>
</tr>
<tr>
<td style="text-align:left;">
aplatyrhynchos_gene_ensembl
</td>
<td style="text-align:left;">
Mallard genes (ASM874695v1)
</td>
<td style="text-align:left;">
ASM874695v1
</td>
</tr>
<tr>
<td style="text-align:left;">
apolyacanthus_gene_ensembl
</td>
<td style="text-align:left;">
Spiny chromis genes (ASM210954v1)
</td>
<td style="text-align:left;">
ASM210954v1
</td>
</tr>
<tr>
<td style="text-align:left;">
applatyrhynchos_gene_ensembl
</td>
<td style="text-align:left;">
Duck genes (CAU_duck1.0)
</td>
<td style="text-align:left;">
CAU_duck1.0
</td>
</tr>
<tr>
<td style="text-align:left;">
atestudineus_gene_ensembl
</td>
<td style="text-align:left;">
Climbing perch genes (fAnaTes1.2)
</td>
<td style="text-align:left;">
fAnaTes1.2
</td>
</tr>
<tr>
<td style="text-align:left;">
bbbison_gene_ensembl
</td>
<td style="text-align:left;">
American bison genes (Bison_UMD1.0)
</td>
<td style="text-align:left;">
Bison_UMD1.0
</td>
</tr>
<tr>
<td style="text-align:left;">
bgrunniens_gene_ensembl
</td>
<td style="text-align:left;">
Domestic yak genes (LU_Bosgru_v3.0)
</td>
<td style="text-align:left;">
LU_Bosgru_v3.0
</td>
</tr>
<tr>
<td style="text-align:left;">
bihybrid_gene_ensembl
</td>
<td style="text-align:left;">
Hybrid - Bos Indicus genes (UOA_Brahman_1)
</td>
<td style="text-align:left;">
UOA_Brahman_1
</td>
</tr>
<tr>
<td style="text-align:left;">
bmusculus_gene_ensembl
</td>
<td style="text-align:left;">
Blue whale genes (mBalMus1.v2)
</td>
<td style="text-align:left;">
mBalMus1.v2
</td>
</tr>
<tr>
<td style="text-align:left;">
bmutus_gene_ensembl
</td>
<td style="text-align:left;">
Wild yak genes (BosGru_v2.0)
</td>
<td style="text-align:left;">
BosGru_v2.0
</td>
</tr>
<tr>
<td style="text-align:left;">
bsplendens_gene_ensembl
</td>
<td style="text-align:left;">
Siamese fighting fish genes (fBetSpl5.2)
</td>
<td style="text-align:left;">
fBetSpl5.2
</td>
</tr>
</tbody>
</table>
<p>Lots of interesting species are among the first top 20, difficult to choose which one to focus on. Cows have interesting things to <a href="https://github.com/tnalpgge/rank-amateur-cowsay">say</a> so lets pick the database of the cow’s genome (<code>btaurus_gene_ensembl</code>) for the following example.</p>
<p>A transcript is the transcribed part of a gene. By transcribed we mean that the genomic region will be used to produce a functional biomolecule (RNA which can then be used to direct the construction of a protein). A transcript can be composed by one or more exons.</p>
<p>The information we are interested in are transcripts and exons. Transcripts are uniquely identified by the attribute <code>ensembl_transcript_id</code> and exons by the attribute <code>ensembl_exon_id</code>. We also need the size of exons (in base pairs) and this we can calculate from the difference in the genomic coordinates of the start and end of each exon stored in the attributes <code>exon_chrom_start</code> and <code>exon_chrom_end</code> respectively. Finally we can reduce the noise in the data we download by excluding non-constitutive exons. To keep the amount of data manageable we also filter by a single chromosome:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ensembl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"ensembl"</span>, dataset <span class="op">=</span> <span class="st">"btaurus_gene_ensembl"</span><span class="op">)</span>
<span class="va">cow_exons</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>filters <span class="op">=</span> <span class="st">'chromosome_name'</span>,
                    values <span class="op">=</span> <span class="fl">13</span>,
                    attributes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl_transcript_id"</span>,
                                 <span class="st">"ensembl_exon_id"</span>,
                                 <span class="st">"exon_chrom_start"</span>,
                                 <span class="st">"exon_chrom_end"</span>,
                                 <span class="st">"is_constitutive"</span><span class="op">)</span>, mart<span class="op">=</span><span class="va">ensembl</span><span class="op">)</span></code></pre></div>
<p>There are 14323 in our table, quite enough to get a statistical signal, with the drawback that we focus on a single chromosome which might have something particular.</p>
<p>To make sure that we did correctly downloaded the information we need we could have added a few more attributes and perform some descriptive statistics or data cleaning but for the purpose of this examples we can skip such steps.</p>
<p>Having downloaded the necessary information we can now process them to create a menzerath object to store the information:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">processed_exons</span> <span class="op">&lt;-</span><span class="op">(</span><span class="va">cow_exons</span> <span class="op">%&gt;%</span> 
                     <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">is_constitutive</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                     <span class="fu">mutate</span><span class="op">(</span>exon_size <span class="op">=</span> <span class="va">exon_chrom_end</span><span class="op">-</span><span class="va">exon_chrom_start</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                     <span class="fu">group_by</span><span class="op">(</span><span class="va">ensembl_transcript_id</span><span class="op">)</span> <span class="op">%&gt;%</span>
                     <span class="fu">summarize</span><span class="op">(</span>transcript_size <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span>, average_exon_size<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">exon_size</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">cow_menzerath</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/menzerath.html">menzerath</a></span><span class="op">(</span><span class="va">processed_exons</span>, x <span class="op">=</span> <span class="st">"transcript_size"</span>, y <span class="op">=</span> <span class="st">"average_exon_size"</span><span class="op">)</span></code></pre></div>
<p>To get a glimpse of the information in our object we can print it:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="va">cow_menzerath</span><span class="op">)</span>
<span class="co">#&gt; Observations: 1481</span>
<span class="co">#&gt; y: 297.857142857143,447.8,261.666666666667,158.416666666667,115.380952380952,144.241379310345...</span>
<span class="co">#&gt; x: 7,5,12,12,21,29...</span></code></pre></div>
<p>and plot the data:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">cow_menzerath</span>, fit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="menzerath_biology_files/figure-html/plot_menzerath-1.png" width="700"></p>
<p>By default for the fit we use the Menzerath-Altman formula (for alternatives see section Alternative models of the vignette Introduction to package menzerath):</p>
<p><span class="math display">\[ y = a \cdot x^b \cdot e^{-cx} \]</span></p>
<p>We could explore a bit more in depth the data by trying a different formula for fitting:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">alt_fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html">fit</a></span><span class="op">(</span><span class="va">cow_menzerath</span>, method<span class="op">=</span><span class="st">"Milicka_2"</span><span class="op">)</span>
<span class="fu"><a href="../reference/get_parameters.html">get_parameters</a></span><span class="op">(</span><span class="va">alt_fit</span><span class="op">)</span>
<span class="co">#&gt; $a</span>
<span class="co">#&gt; [1] 271.7336</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $b</span>
<span class="co">#&gt; [1] 0.1831749</span></code></pre></div>
<p>Interpreting the meaning of the parameters in most cases is not straightforwards even in cases where the system is well known. To get a better understanding of Menzerath’s law in the particular system one could also compare across systems. To make the process less repetitive we can define a helper function <code>ensemble2menzerath</code> to download the necessary information (see <a href="#Useful-functions">Useful functions</a> for the full code).</p>
<p>As the following code is too heavy and time consuming as it requests and processes a lot of information, it has been pre-run and the resulting plot is directly inserted below:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">yms</span> <span class="op">&lt;-</span> <span class="fu">ensemble2menzerath</span><span class="op">(</span><span class="op">)</span>
<span class="va">hms</span> <span class="op">&lt;-</span> <span class="fu">ensemble2menzerath</span><span class="op">(</span><span class="st">"hsapiens_gene_ensembl"</span><span class="op">)</span>
<span class="va">mms</span> <span class="op">&lt;-</span> <span class="fu">ensemble2menzerath</span><span class="op">(</span><span class="st">"mmusculus_gene_ensembl"</span><span class="op">)</span>
<span class="va">cms</span> <span class="op">&lt;-</span> <span class="fu">ensemble2menzerath</span><span class="op">(</span><span class="st">"btaurus_gene_ensembl"</span><span class="op">)</span>
<span class="va">yeast_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">yms</span>, fit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">human_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">hms</span>, fit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">mouse_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">mms</span>, fit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="va">cow_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="va">cms</span>, fit <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">yeast_plot</span>, <span class="va">human_plot</span>, <span class="va">mouse_plot</span>, <span class="va">cow_plot</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th>species</th>
<th>scientific name</th>
<th>database name</th>
<th>number of observations</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>yeast</td>
<td><em>S. cerevisiae</em></td>
<td><code>scerevisiae_gene_ensembl</code></td>
<td>7,127</td>
</tr>
<tr class="even">
<td>human</td>
<td><em>H. sapiens</em></td>
<td><code>hsapiens_gene_ensembl</code></td>
<td>52,351</td>
</tr>
<tr class="odd">
<td>mouse</td>
<td><em>M. musculus</em></td>
<td><code>mmusculus_gene_ensembl</code></td>
<td>43,864</td>
</tr>
<tr class="even">
<td>cow</td>
<td><em>B. taurus</em></td>
<td><code>btaurus_gene_ensembl</code></td>
<td>42,022</td>
</tr>
</tbody>
</table>
<div class="figure">
<img src="menzerath_biology_files/exons_menzerath.png" alt=""><p class="caption">Menzerath’s law in multiple genomes</p>
</div>
<div id="troubleshooting" class="section level2">
<h2 class="hasAnchor">
<a href="#troubleshooting" class="anchor"></a>Troubleshooting</h2>
<div id="Install-missing-libraries-section" class="section level3">
<h3 class="hasAnchor">
<a href="#Install-missing-libraries-section" class="anchor"></a>Install Missing Libraries</h3>
<p>If <a href="https://www.tidyverse.org/">tidyverse</a> or <a href="https://wilkelab.org/cowplot/index.html">cowplot</a> is missing they can be installed with:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw">install.packages</span>(c<span class="st">"tidyverse"</span>, <span class="st">"cowplot"</span>)</span></code></pre></div>
<p>If <a href="https://sellisd.github.io/menzerath/">menzerath</a> is missing:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"sellisd/menzerath"</span>, build_vignettes<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>If <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> is it can be installed with the following commands (this might take some time):</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"biomaRt"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="Useful-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#Useful-functions" class="anchor"></a>Useful functions</h2>
<p>The full code for the helper function is the following:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">#' Generate a menzerath object from exons</span>
<span class="co">#'</span>
<span class="co">#' @param dataset the name of an ensemble dataset by default one for _S. cerevisiae_</span>
<span class="co">#'</span>
<span class="co">#' @return</span>
<span class="co">#' @export</span>
<span class="co">#'</span>
<span class="co">#' @examples</span>
<span class="va">ensemble2menzerath</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">dataset</span> <span class="op">=</span> <span class="st">"scerevisiae_gene_ensembl"</span>, <span class="va">...</span><span class="op">)</span><span class="op">{</span>
  <span class="va">ensembl</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"ensembl"</span>, dataset <span class="op">=</span> <span class="va">dataset</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"Downloading..."</span><span class="op">)</span>
  <span class="va">exons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl_transcript_id"</span>,
                              <span class="st">"transcript_length"</span>,
                              <span class="st">"ensembl_exon_id"</span>,
                              <span class="st">"exon_chrom_start"</span>,
                              <span class="st">"exon_chrom_end"</span>,
                              <span class="st">"is_constitutive"</span>, <span class="va">...</span><span class="op">)</span>, mart<span class="op">=</span><span class="va">ensembl</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="st">"done"</span><span class="op">)</span>
  <span class="va">processed</span> <span class="op">&lt;-</span><span class="op">(</span><span class="va">exons</span> <span class="op">%&gt;%</span> 
                 <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">is_constitutive</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                 <span class="fu">mutate</span><span class="op">(</span>exon_size <span class="op">=</span> <span class="va">exon_chrom_end</span><span class="op">-</span><span class="va">exon_chrom_start</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                 <span class="fu">group_by</span><span class="op">(</span><span class="va">ensembl_transcript_id</span><span class="op">)</span> <span class="op">%&gt;%</span>
                 <span class="fu">summarize</span><span class="op">(</span>transcript_size <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span>, average_exon_size<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">exon_size</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
  
  <span class="va">ms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/menzerath.html">menzerath</a></span><span class="op">(</span><span class="va">processed</span>, x <span class="op">=</span> <span class="st">"transcript_size"</span>, y <span class="op">=</span> <span class="st">"average_exon_size"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Baixeries2013">
<p>Baixeries, Jaume, Antoni Hernández-Fernández, Núria Forns, and Ramon Ferrer-i-Cancho. 2013. “The Parameters of the Menzerath-Altmann Law in Genomes.” <em>Journal of Quantitative Linguistics</em> 20 (2): 94–104. <a href="https://doi.org/10.1080/09296174.2013.773141">https://doi.org/10.1080/09296174.2013.773141</a>.</p>
</div>
<div id="ref-FerrerICancho2009">
<p>Ferrer-I-Cancho, Ramon, and NÃria Forns. 2009. “The Self-Organization of Genomes.” <em>Complexity</em>, NA–NA. <a href="https://doi.org/10.1002/cplx.20296">https://doi.org/10.1002/cplx.20296</a>.</p>
</div>
<div id="ref-Li2011">
<p>Li, Wentian. 2011. “Menzeraths Law at the Gene-Exon Level in the Human Genome.” <em>Complexity</em> 17 (4): 49–53. <a href="https://doi.org/10.1002/cplx.20398">https://doi.org/10.1002/cplx.20398</a>.</p>
</div>
<div id="ref-Nikolaou2014">
<p>Nikolaou, Christoforos. 2014. “Menzerath–Altmann Law in Mammalian Exons Reflects the Dynamics of Gene Structure Evolution.” <em>Computational Biology and Chemistry</em> 53: 134–43. <a href="https://doi.org/https://doi.org/10.1016/j.compbiolchem.2014.08.018">https://doi.org/https://doi.org/10.1016/j.compbiolchem.2014.08.018</a>.</p>
</div>
<div id="ref-Shahzad2015">
<p>Shahzad, Khuram, Jay E. Mittenthal, and Gustavo Caetano-Anollés. 2015. “The Organization of Domains in Proteins Obeys Menzerath-Altmann’s Law of Language.” <em>BMC Systems Biology</em> 9 (1). <a href="https://doi.org/10.1186/s12918-015-0192-9">https://doi.org/10.1186/s12918-015-0192-9</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Diamantis Sellis.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
